### PR TITLE
[TrimmableTypeMap] Detect overrides and constructors in JavaPeerScanner

### PR DESCRIPTION
**Pull request by simonrozsival**
## TL;DR

`JavaPeerScanner` now matches legacy `CecilImporter` output for method overrides, constructor chaining, and interface implementations. The scanner uses a 5-pass approach: direct `[Register]`, property `[Register]`, base class override detection, interface method detection, and constructor chain detection. Also fixes 4 bugs in the integration test harness that were hiding mismatches between legacy and new scanner output. Validated with zero divergence across ~8000 Mono.Android types — no filtering, direct 1:1 comparison.

247 unit tests + 10 integration tests pass.

Part of #10789

## Context

The `JavaPeerScanner` was missing several categories of methods that the legacy `CecilImporter` finds:

1. **Method overrides without `[Register]`** — User types (e.g., `MainActivity`) override `Activity.OnCreate()` without `[Register]` on the override. The attribute is only on the base class method in Mono.Android.

2. **Java constructors via base ctor chaining** — User types have non-activation constructors without `[Register]`. The legacy `CecilImporter` walks the base type hierarchy, seeds registered ctors, and accepts derived ctors with compatible parameters.

3. **Constructor parameter JNI signatures for object types** — Ctors taking Java peer types (e.g., `X509Certificate[]`) need `L<jniName>;` descriptors, not just primitive mappings.

4. **Base JNI name for types without `[Register]`** — Types like `XmlReaderPullParser` have no `[Register]` but are Java peers with auto-computed JNI names. `ResolveBaseJavaName` needs to compute these on demand.

## Scanner architecture

The scanner's `CollectMarshalMethods` now uses a 5-pass approach:

```
CollectMarshalMethods(typeDef, detectBaseOverrides)
│
├─ Pass 1: Direct [Register] / [Export] / [ExportField] on methods
├─ Pass 2: [Register] on properties (getter)
│
│  (Passes 3-5 only for user ACW types: detectBaseOverrides=true)
│
├─ Pass 3: Override detection — walk base hierarchy
│  ├─ Method overrides (FindBaseRegisteredMethodInfo)
│  └─ Property overrides (FindBaseRegisteredProperty)
├─ Pass 4: Interface method implementations
│  └─ CollectInterfaceMethodImplementations
└─ Pass 5: Constructor chaining from base registered ctors
   └─ CollectBaseConstructorChain
      ├─ Seed base registered ctors
      ├─ Match params → already covered
      └─ Fallback → parameterless super() + BuildJniCtorSignature
```

### Scanner (`JavaPeerScanner.cs`, `JavaPeerInfo.cs`)

**Pass 3 — Override detection:** For each virtual override without `[Register]`, walks the base hierarchy to find the registered base method and copies its registration info. Also handles property overrides (e.g., `Throwable.Message`). Populates `DeclaringTypeName`/`DeclaringAssemblyName` so UCO wrappers call `n_*` on the correct base type. Walks all base types with no DoNotGenerateAcw boundary — matching legacy `GetBaseDefinition`/`GetBaseRegisteredMethod`.

**Pass 5 — Constructor chain detection:** Mirrors legacy `CecilImporter` behavior:
- Seeds all base registered ctors directly into the type's marshal methods
- Matches derived ctors with compatible parameters against base registered ctors
- Fallback: if a base has a registered `()V` ctor, accepts derived ctors with novel params (generates `super()` call via `SuperArgumentsString=""`)
- Rejects ctors with non-Java parameter types (`System.IntPtr`, `System.Action`, etc.)
- Resolves Java peer object types (e.g., `X509Certificate`) to `L<jniName>;` via assembly cache lookup
- Handles `[JniConstructorSignature]` (Java.Interop-style) in addition to `[Register]`
- Stops base walk at `DoNotGenerateAcw` boundary (matching legacy)

**`ResolveBaseJavaName`** — Now computes auto JNI names for base types without `[Register]` when they haven't been scanned yet (scan order within an assembly is not guaranteed).

**`ManagedTypeToJniDescriptorOrNull`** — Now resolves Java peer object types by looking up their `[Register]` JNI name in the assembly cache.

### Unit tests + fixtures

- **`OverrideDetectionTests`** (10 tests) — overrides without `[Register]`, deep inheritance (3 levels), ACW→ACW→MCW hierarchy, property overrides, mixed methods, new-slot exclusion, DeclaringTypeName verification
- **`ConstructorDetectionTests`** (9 tests) — base ctor chain, implicit default ctor, base ctor seed, `JniConstructorSignature`, parameterless fallback, same-arity type mismatch, multi-parameter signature computation
- Test fixtures: `UserActivity`, `FullActivity`, `DeeplyDerived`, `MixedMethods`, `NewSlotActivity`, `CustomException`, `BaseFragment`, `DerivedFragment`, `GrandchildFragment`, `JiStyleView`, `ActivityWithCustomCtor`, `DialogBase`, `CustomDialog`, `ActivityWithMultiParamCtor`

---

### Integration test harness improvements

_The following changes are not directly related to the scanner features above, but were necessary to properly validate them. The scanner now produces more complete output (interface methods, constructors with object type params, types without `[Register]`), which exposed pre-existing bugs in how the legacy test harness transforms `CecilImporter` output for comparison._

**`GetCecilJavaName`** — Added `JavaNativeTypeManager.ToJniName` fallback for types without `[Register]` (e.g., `ActivityTracker`). Previously returned `null`, silently skipping these types. Now matches what `CecilImporter.CreateType` actually does.

**`ExtractMethodRegistrations`** — Fixed to use `m.JavaName` (JNI name) instead of `m.Name` (managed name) for `[Export]` methods. Now includes constructors from `wrapper.Constructors` so both scanners' constructors are compared directly.

**Removed all filtering** — No more interface method filtering, no phantom entry fallback, no constructor exclusion. Both `RunLegacy` and `RunNew` produce complete unfiltered output compared 1:1 across ~8000 Mono.Android types.

---

### Related issue

Filed #10931 — `JcwJavaSourceGenerator` must skip `registerNatives` in the static block for Application/Instrumentation types.

## Test results

- 247 unit tests pass
- 10 integration tests pass (including `ExactMarshalMethods_MonoAndroid` and `ExactJavaConstructors_MonoAndroid` — no filtering, direct 1:1 comparison)

## Legacy comparison notes

- **Override detection** matches legacy `GetBaseDefinition` → `GetBaseRegisteredMethod`: walks all base types (no DoNotGenerateAcw boundary), checks `Virtual || Abstract`, matches by name + parameter compatibility.
- **Constructor chaining** matches legacy `ctorTypes` loop (CecilImporter.cs:124-160): stops at DoNotGenerateAcw, seeds registered base ctors, checks `AreParametersCompatibleWith`, falls back to parameterless `super()`.
- **Object type resolution** matches legacy `GetJniTypeName`: resolves Java peer types to `L<jniName>;` via `[Register]` lookup, returns null for non-Java types.
- **Parameter compatibility** uses string comparison of decoded S.R.M. signatures — equivalent to Cecil's `a.FullName == b.FullName` for non-generic cases. Validated against ~8000 types with zero divergence.

### CODE REVIEW

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs [hunk 1]
```diff
@@ -264,12 +269,478 @@ List<MarshalMethodInfo> CollectMarshalMethods (TypeDefinition typeDef, AssemblyI
...
+
+			var getterName = index.Reader.GetString (getterDef.Name);
+			if (alreadyRegistered.Contains (getterName)) {
+				continue;
+			}
+
+			var baseRegistration = FindBaseRegisteredProperty (typeDef, index, getterName, getterDef);
+			if (baseRegistration is not null) {
+				methods.Add (baseRegistration);
+				alreadyRegistered.Add (getterName);
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs]: Fixed in c951c5b51 — `CollectBasePropertyOverrides` now uses the same `"$"{getterName}({string.Join (",", sig.ParameterTypes)})"` key format as Pass 1/2, so directly-registered property getters are correctly recognized and won't produce duplicates.

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs [hunk 2]
```diff
@@ -264,12 +269,478 @@ List<MarshalMethodInfo> CollectMarshalMethods (TypeDefinition typeDef, AssemblyI
...
+			if (!AreParametersCompatible (derivedMethod, index, baseMethodDef, baseIndex)) {
+				continue;
+			}
+
+			// Found a matching base method — check if it has [Register]
+			if (TryGetMethodRegisterInfo (baseMethodDef, baseIndex, out var registerInfo, out _) && registerInfo is not null) {
+				return registerInfo;
+			}
+		}
+
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs:751]: Fixed in c951c5b51 — `FindBaseRegisteredMethodInfo` now checks the `DoNotGenerateAcw` boundary before recursing, matching the pattern used in `CollectBaseRegisteredCtors`. If the base type has `DoNotGenerateAcw=true`, we inspect its methods but don't recurse further up.

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs [hunk 3]
```diff
@@ -264,12 +269,478 @@ List<MarshalMethodInfo> CollectMarshalMethods (TypeDefinition typeDef, AssemblyI
...
+					ManagedMethodName = getterName,
+					NativeCallbackName = $"n_{getterName}",
+					IsConstructor = false,
+				};
+			}
+		}
+
+		// Recurse up
+		return FindBaseRegisteredProperty (baseTypeDef, baseIndex, getterName, derivedGetter);
+	}
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs:832]: Fixed in c951c5b51 — same DoNotGenerateAcw boundary check added to `FindBaseRegisteredProperty`, consistent with `FindBaseRegisteredMethodInfo` and `CollectBaseRegisteredCtors`.

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/TestFixtures/TestTypes.cs [hunk 4]
```diff
@@ -308,6 +309,122 @@ public class UnregisteredExporter : Java.Lang.Object
...
+	[Register ("my/app/JiStyleView")]
+	public class JiStyleView : Android.Views.View
+	{
+		protected JiStyleView (IntPtr handle, JniHandleOwnership transfer) : base (handle, transfer) { }
+
+		[Java.Interop.JniConstructorSignature ("()V")]
+		public JiStyleView () : base (default!, default) { }
+
+		[Java.Interop.JniConstructorSignature ("(Landroid/content/Context;)V")]
+		public JiStyleView (Context context) : base (default!, default) { }
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/TestFixtures/TestTypes.cs]: Fixed — replaced default! with default. Both parameters (IntPtr and JniHandleOwnership) are value types so the null-forgiving operator was unnecessary.

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.IntegrationTests/TypeDataBuilder.cs [hunk 5]
```diff
@@ -96,22 +97,26 @@ public static (Dictionary<string, TypeComparisonData> perType, List<TypeMapEntry
...
+				try {
+					var wrapper = CecilImporter.CreateType (typeDef, cache);
+					foreach (var ctor in wrapper.Constructors) {
+						if (!string.IsNullOrEmpty (ctor.JniSignature)) {
+							javaCtorSignatures.Add (ctor.JniSignature);
 						}
 					}
+				} catch {
+					// Some types may fail CecilImporter (e.g., missing base types).
+					// Fall back to direct attribute scanning.
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply tests/Microsoft.Android.Sdk.TrimmableTypeMap.IntegrationTests/TypeDataBuilder.cs]: Fixed — removed the try/catch entirely. If CecilImporter.CreateType throws for a type, the test should surface that failure rather than silently falling back to different extraction logic. We're matching legacy behavior, not extending it.

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict:** ✅ LGTM

Found **0 issues** in the updated diff.

The previous round of review feedback has been fully addressed:
- `Signature!` null-forgiving operators replaced with proper null guards
- `FindBaseRegisteredMethodInfo` and `FindBaseRegisteredProperty` now stop at `DoNotGenerateAcw` boundary
- Property override key format made consistent with Pass 1/2
- `default!` replaced with `default` in test fixtures
- Bare `catch` in TypeDataBuilder removed
- Duplicate `if (detectBaseOverrides)` blocks merged
- Override detection tests consolidated (14→7 Facts, all assertions preserved)

👍 The override detection and constructor chaining logic is well-structured, thoroughly documented, and correctly mirrors legacy `CecilImporter` behavior. The 4-pass architecture (direct [Register], property [Register], base overrides, base ctor chain) has clear separation of concerns. Test coverage is comprehensive across edge cases (deep inheritance, new-slot exclusion, property overrides, JI-style attributes, parameterless fallback).

⚠️ CI is still pending (build 1334063) — cannot confirm merge-readiness until checks pass.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.IntegrationTests/TypeDataBuilder.cs [hunk 6]
```diff
@@ -96,22 +97,25 @@ public static (Dictionary<string, TypeComparisonData> perType, List<TypeMapEntry
...
+			if (!typeDef.IsInterface && !ScannerRunner.HasDoNotGenerateAcw (typeDef)) {
+				try {
+					var wrapper = CecilImporter.CreateType (typeDef, cache);
+					foreach (var ctor in wrapper.Constructors) {
+						if (!string.IsNullOrEmpty (ctor.JniSignature)) {
+							javaCtorSignatures.Add (ctor.JniSignature);
 						}
 					}
+				} catch (Exception ex) {
+					System.Diagnostics.Debug.WriteLine ($"CecilImporter.CreateType failed for {typeDef.FullName}: {ex.Message}");
```
Comment (jonathanpeppers) [inline tests/Microsoft.Android.Sdk.TrimmableTypeMap.IntegrationTests/TypeDataBuilder.cs]: Does this code have some kind of logging infrastructure?

We would need MSBuild tasks to be able to pass in a delegate for logging, similar to:

https://github.com/dotnet/android/blob/ff3960706402b1f9af95874c493936c119738c08/build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/JdkInfo.cs#L89-L105

We can file an issue to do this in the future.

Comment (simonrozsival) [reply tests/Microsoft.Android.Sdk.TrimmableTypeMap.IntegrationTests/TypeDataBuilder.cs]: I think this log is probably unnecessary going forward. I will remove it.

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict:** ✅ LGTM

Found **0 issues** across 14 changed files (~1800 additions).

No null-forgiving operators, no empty catch blocks, no filtering or hidden data in scanner comparison. Both `RunLegacy` and `RunNew` produce complete unfiltered output (methods, properties, interface implementations, constructors) compared 1:1 across ~8000 Mono.Android types.

The 5-pass scanner architecture is well-structured:
- Pass 1-2: direct `[Register]`/`[Export]` on methods and properties
- Pass 3: override detection walking base hierarchy (no DoNotGenerateAcw boundary, matching legacy `GetBaseDefinition`)
- Pass 4: interface method implementations
- Pass 5: constructor chaining with parameterless fallback

The integration test harness fixes (4 bugs) are clearly separated in commits and correctly documented as supporting infrastructure.

⚠️ CI: Linux ✅, Windows/Mac still pending. Cannot confirm merge-readiness until all checks pass.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict:** ✅ LGTM

Found **0 issues** across 14 changed files (~2400 line diff).

No null-forgiving operators, no empty catch blocks, no filtering or hidden data in the comparison. Code duplication has been addressed:
- `TryResolveBaseType` extended to share base type resolution across 3 methods
- `TryGetPrimitiveJniDescriptor` extracted to share JNI mapping between nullable and non-nullable variants
- `AddMarshalMethod` reused in `CollectInterfaceMethodImplementations` (eliminated ~20 lines of duplicated `MarshalMethodInfo` construction)
- `HaveIdenticalParameterTypes` renamed from `AreParametersCompatible` with unused params removed

Test coverage is comprehensive: 254 unit tests + 10 integration tests. Recent additions cover abstract base overrides, correct overload selection, Java peer object arrays in ctor signatures, unsigned primitives, empty connectors, interface inheritance chains, and ACW→ACW `DeclaringTypeName` resolution.

⚠️ CI is pending (build 1337757). Previous build (1337248) was all green on Linux/Mac/Windows.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Review by jonathanpeppers — APPROVED** [review]
